Add support to log to redis

ruudvdd opened the pull request:
Updated version of https://github.com/symfony/monolog-bundle/pull/262

ruudvdd commented:
Got the same deprecation notice as https://github.com/symfony/monolog-bundle/pull/279 and can't find the reason other than it's a known issue in older versions of phpunit.

Ref: https://github.com/symfony/maker-bundle/issues/330

lyrixx commented on DependencyInjection/Configuration.php:
```diff
@@ -517,6 +529,44 @@ public function getConfigTreeBuilder()
+                                    ->ifTrue(function ($v) {
+                                        return !isset($v['id']) && !isset($v['host']);
+                                    })
+                                    ->thenInvalid('What must be set is either the host or the id.')
```
Could you be more explicit about what is an `id` ?
Thanks

lyrixx commented on DependencyInjection/Configuration.php:
```diff
@@ -517,6 +529,44 @@ public function getConfigTreeBuilder()
+                                    ->ifTrue(function ($v) {
+                                        return !isset($v['id']) && !isset($v['host']);
+                                    })
+                                    ->thenInvalid('What must be set is either the host or the id.')
```
same here.

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -291,6 +291,43 @@ private function buildHandler(ContainerBuilder $container, $name, array $handler
             ));
             break;
+        case 'redis':
+        case 'predis':
```
I think this does not work, since you are using `$handler['redis]` bellow (and not `$handler['predis]`)

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -291,6 +291,43 @@ private function buildHandler(ContainerBuilder $container, $name, array $handler
+            if (isset($handler['redis']['id'])) {
+                $clientId = $handler['redis']['id'];
+            } elseif ('redis' === $handler['type']) {
+                if (class_exists('\Redis')) {
```
` if (!class_exists('\Redis')) { throw...}` (same after)

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -801,6 +838,9 @@ private function getHandlerClassByType($handlerType)
             'server_log' => 'Symfony\Bridge\Monolog\Handler\ServerLogHandler',
+            'redis' => 'Monolog\Handler\RedisHandler',
+            'predis' => 'Monolog\Handler\RedisHandler',
+
```
extra line

lyrixx commented:
I also forgot to ask you to add a note in the [CHANGELOG](https://github.com/symfony/monolog-bundle/blob/master/CHANGELOG.md#332-2018-12-29) in v3.3.2

lyrixx commented:
Also, could you rebase please? thanks

ruudvdd commented:
@lyrixx 
Thanks for reviewing. I've fixed your comments!

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -291,6 +291,45 @@ private function buildHandler(ContainerBuilder $container, $name, array $handler
+                $clientId = $handler[$type]['id'];
+            } elseif ('redis' === $type) {
+                if (!class_exists('\Redis')) {
+                    throw new \RuntimeException('The \Redis is not available.');
```
I think the following is better is better:
> The Redis  **class** is not available

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -291,6 +291,45 @@ private function buildHandler(ContainerBuilder $container, $name, array $handler
+                $container->setDefinition($clientId, $client);
+            } else {
+                if (!class_exists('\Predis\Client')) {
+                    throw new \RuntimeException('The \Predis\Client is not available.');
```
Add class here too

lyrixx commented on DependencyInjection/MonologExtension.php:
```diff
@@ -291,6 +291,45 @@ private function buildHandler(ContainerBuilder $container, $name, array $handler
+            if (isset($handler[$type]['id'])) {
+                $clientId = $handler[$type]['id'];
+            } elseif ('redis' === $type) {
+                if (!class_exists('\Redis')) {
```
We don't prefix the FQCN with a `\`, it should be removed

lyrixx commented:
Oh, you forgot to update the README.md file. Thanks

lyrixx commented:
Hello, It looks like something goes wrong with git.
Could you rebase please?
Thanks

ruudvdd commented:
I did a rebase before applying my changes, but I did it again.

I've changed the configuration documentation comment a bit, because it wasn't correct. Also, I reverted the $type change in the MonologExtension because it was correct the first time. It just wasn't documented right. 'redis' is the property in the handler config where the redis host, id, ... are stored. They're both named 'redis' because it's a reference to the technology, not the lib.

Also applied the fixes for your above comments

lyrixx commented:
I think you did a merge, and not a rebase. I will try to fix it for you :)

lyrixx commented:
Thanks for your hardwork.
I have finished it in #305, I have rebased your commit and squashed theme into one.
